engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java

```diff
@@ -142,14 +142,31 @@ private Object get(int index, Node node) {
     }
   }
 
+  private static boolean checkArraySize(boolean mustBeExact, int real, int expected) {
+    if (real == expected) {
+      return true;
+    } else {
+      if (mustBeExact) {
+        CompilerDirectives.transferToInterpreter();
+      }
+      return false;
+    }
+  }
+
   /** Returns the current array of the builder. */
-  private Object toArray() {
+  private Object toArray(boolean mustBeExact) {
     if (objectArray != null) {
-      return objectArray.length == size ? objectArray : Arrays.copyOf(objectArray, size);
+      return checkArraySize(mustBeExact, objectArray.length, size)
```

Comment:
The graph we get is **too wide**:

![too wide IGV](https://github.com/user-attachments/assets/c34ab9ce-20b0-4643-80ad-a6b64c5c1339)

Ideal graph should _only contain branching when the user code does_ some `if` decisions. That's not the case right now.

9823d2ceae3472ec4227e3e06f876f3eb02d5e02 is first step towards that direction. In many cases (e.g.  `vectorFromAFunction`) we know the array is properly pre-allocated. We don't want `Arrays.copyOf` in the graph at all.